app-creator/client/src/utils/helpers.ts

```diff
@@ -22,7 +33,127 @@ export function getIdPrefixLastIndex(id: string) {
   return id.slice(0, id.lastIndexOf('-'));
 }
 
+/**
+ * Generates random ids with length 32.
+ */
+export function generateRandomId() {
+  return [...Array(32)]
+    .map((_) => (~~(Math.random() * 36)).toString(36))
+    .join('');
+}
+
 /**
  * Empty function. Used as a placeholder.
  */
 export const noop: VoidFunction = () => {};
+
+/**
+ * Triggered everytime we want to render a new template in the storyboard.
+ */
+export function generateUI(
+  template: AppCreatorStore['template'],
+  node: HTMLElement
+) {
+  const templateCopy = Object.assign({}, template);
+
+  function helper(widgetData: WidgetMetaData): HTMLElement {
+    const { id, children } = widgetData;
+    const { element, dropzone, map, draggable } = getWidgetElement(widgetData);
+
+    for (const childID of children) {
+      if (dropzone != null) {
+        dropzone.appendChild(helper(templateCopy[childID]));
+      } else {
+        element.appendChild(helper(templateCopy[childID]));
+      }
+    }
+
+    if (map != null) {
+      templateCopy[id].widgetRef = map;
+    } else {
+      templateCopy[id].widgetRef = element;
+    }
+
+    return draggable == null ? element : draggable;
+  }
+
+  // The root of the template will always have an id of panel-template-0
+  const root = templateCopy[ROOT_ID];
+  node.appendChild(helper(root));
+
+  // Replace the store's template with the one that include the widgetRefs.
+  store.dispatch(setSelectedTemplate(templateCopy));
+}
+
+/**
+ * Helper function to create ui widgets.
+ */
+export function getWidgetElement({
+  id,
+  editable,
+  uniqueAttributes,
+  style,
+}: WidgetMetaData): {
+  element: HTMLElement;
+  dropzone: Dropzone | null;
+  map: Map | null;
+  draggable: DraggableWidget | null;
+} {
+  // Get widget type (i.e. panel-0 -> panel).
+  const type = getIdPrefix(id);
+
+  // Create DOM element.
+  let element = document.createElement(`ui-${type}`);
+  element.id = id;
+
+  // Set Unique attributes.
+  for (const attribute in uniqueAttributes) {
+    element.setAttribute(attribute, uniqueAttributes[attribute]);
+  }
+
+  // Set styles.
+  (element as EEWidget).setStyle(style);
```

Comment:
Is element guaranteed to be have setStyle()? If not, make an assertion like `if (element.setStyle)` first